presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -0,0 +1,109 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hudi.HudiSessionProperties.getSplitGeneratorParallelism;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A runnable to load Hudi splits asynchronously in the background
+ */
+public class HudiBackgroundSplitLoader
+        implements Runnable
+{
+    private static final Logger LOG = Logger.get(HudiBackgroundSplitLoader.class);
+    private final ConnectorSession session;
+    private final ExtendedHiveMetastore metastore;
+    private final HudiTableLayoutHandle layout;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final List<String> partitions;
+    private final String latestInstant;
+    private final int splitGeneratorNumThreads;
+    private final ExecutorService splitGeneratorExecutorService;
+
+    public HudiBackgroundSplitLoader(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            List<String> partitions,
+            String latestInstant)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.layout = requireNonNull(layout, "layout is null");
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.partitions = requireNonNull(partitions, "partitions is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+
+        this.splitGeneratorNumThreads = getSplitGeneratorParallelism(session);
+        this.splitGeneratorExecutorService = Executors.newFixedThreadPool(splitGeneratorNumThreads);
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+        List<HudiPartitionSplitGenerator> splitGeneratorList = new ArrayList<>();
+        List<Future> splitGeneratorFutures = new ArrayList<>();
+        ArrayDeque<String> partitionQueue = new ArrayDeque<>(partitions);
+
+        // Start a number of partition split generators to generate the splits in parallel
+        for (int i = 0; i < splitGeneratorNumThreads; i++) {
+            HudiPartitionSplitGenerator generator = new HudiPartitionSplitGenerator(
+                    session, metastore, layout, fsView, asyncQueue, partitionQueue, latestInstant);
+            splitGeneratorList.add(generator);
+            splitGeneratorFutures.add(splitGeneratorExecutorService.submit(generator));
+        }
+
+        for (HudiPartitionSplitGenerator generator : splitGeneratorList) {
+            // Let the split generator stop once the partition queue is empty
+            generator.stopRunning();
```

Comment:
As how I understand the code, `stopRunning` and `isRunning` have no real control to `HudiPartitionSplitGenerator`

Reply:
The `isRunning` flag is to make sure if `partitionQueue` is somehow empty at the beginning, it still keeps waiting for the partition information to be added to the `partitionQueue` before processing.  Currently it does not have impact on the logic, but I choose to keep it for flexibility in case we want to do async partition listing in the future as well.

Reply:
+1 again. Can we add it later?